### PR TITLE
add InspectContainer and CreateContainerExec in engine.go and called in handler.go

### PR DESCRIPTION
**Pull request by allencloud**
I found with API `GET /containers/{name:.*}/json` and `POST /containers/{name:.*}/exec` in handler.go, Swarm constructs a request by his own to get an response, then manually inserts Node and others in response.

It is costy to insert data and do the replace thing. And this is inconsistency in form data.

This PR did:
1. add InspectContainer and CreateContainerExec in engine.go;
2. in handler, just call InspectContainer and CreateContainerExec;
3. In engine.go add EngineToContainerNode.

Signed-off-by: allencloud allen.sun@daocloud.io

### CODE REVIEW

**Comment by jimmyxian** [comment]
@allencloud 
`InspectContainer` in Swarm and Mesos is all the same. So I think we don't need add a new interface in cluster.go. Just rewrite this in handler.go.

**Comment by allencloud** [comment]
@jimmyxian 
Yes, Totally reasonable. Updated the PR.

**Comment by jimmyxian** [comment]
LGTM

**Comment by allencloud** [comment]
ping @dongluochen @nishanttotla @vieux @abronan

File: api/handlers.go [hunk 1]
```diff
@@ -497,42 +497,24 @@ func getContainerJSON(c *context, w http.ResponseWriter, r *http.Request) {
...
-		return
-	}
-
-	n, err := json.Marshal(container.Engine)
-	if err != nil {
-		httpError(w, err.Error(), http.StatusInternalServerError)
-		return
+	for _, portBindings := range con.NetworkSettings.Ports {
+		for key, portBinding := range portBindings {
+			if portBinding.HostIP == "0.0.0.0" {
```
Comment (dongluochen) [inline api/handlers.go]: Swarm should support all IP address family.

File: api/handlers.go [hunk 2]
```diff
@@ -497,42 +497,24 @@ func getContainerJSON(c *context, w http.ResponseWriter, r *http.Request) {
...
-	if err != nil {
-		httpError(w, err.Error(), http.StatusInternalServerError)
-		return
-	}
-
-	n, err := json.Marshal(container.Engine)
-	if err != nil {
-		httpError(w, err.Error(), http.StatusInternalServerError)
-		return
+	for _, portBindings := range con.NetworkSettings.Ports {
```
Comment (dongluochen) [inline api/handlers.go]: `NetworkSettings` is a pointer and should be validated before use.

File: api/handlers.go (hunk 1 above)
Comment (allencloud) [inline api/handlers.go]: Could you please tell what all IP address family does? @dongluochen 
Actually I can not follow you. Really sorry.

Comment (dongluochen) [inline api/handlers.go]: Swarm should run under IPv4 or IPv6. Here you are trying to check if address is unspecified. Golang has [IsUnspecified](https://golang.org/src/net/ip.go) supporting both IPv4 and IPv6.

**Comment by nishanttotla** [comment]
Ping @allencloud could you please rebase so that CI is green?

**Comment by nishanttotla** [comment]
@allencloud the engine 1.11 test failure probably is related?

**Comment by nishanttotla** [comment]
@allencloud do you want to still pursue this PR? It needs a rebase.

**Comment by allencloud** [comment]
Sorry for the delay, currently I updated the PR, PTAL. @nishanttotla @dongluochen

File: api/handlers.go [hunk 3]
```diff
@@ -570,51 +566,31 @@ func getContainerJSON(c *context, w http.ResponseWriter, r *http.Request) {
...
-			orig = fmt.Sprintf(`"HostIp":"%s"`, net.IPv4zero.String())
-		} else {
-			orig = fmt.Sprintf(`"HostIp":"%s"`, net.IPv6zero.String())
+	con.Node = container.Engine.EngineToContainerNode()
+
+	// update zero ip to engine ip, including IPv4 and IPv6
+	if con.NetworkSettings != nil {
+		for _, portBindings := range con.NetworkSettings.Ports {
+			for key, portBinding := range portBindings {
+				if portBinding.HostIP == net.IPv4zero.String() {
```
Comment (dongluochen) [inline api/handlers.go]: The two `if` may be written without testing IPv4/IPv6. 

```
if ip := net.ParseIP(portBinding.HostIP); ip == nil || ip.IsUnspecified() {
    portBindings[key].HostIP = container.Engine.IP
}
```

Comment (allencloud) [inline api/handlers.go]: Thanks for your feedback. Updated. @dongluochen

**Comment by dongluochen** [comment]
LGTM

**Pull request merged**